riscv: move frametable to data section to improve code relocatability

alexfanqi opened the pull request:
This fixes https://github.com/ocaml/ocaml/issues/10688 and passes all tests on Gentoo riscv64.

I find out the issue is somehow frame descriptor table contains relocations.  A simple example is otherlibs/str. This little pr moves frametable to .data section

`objdump -D -R otherlibs/str.cmxs` shows
```
Disassembly of section .rodata:

000000000000ef98 <caml_shared_startup__frametable>:
    ef98:       0006                    c.slli  zero,0x1
        ...
                        efa0: R_RISCV_RELATIVE  *ABS*+0x9098
    efa6:       0000                    unimp
    efa8:       0012                    c.slli  zero,0x4
    efaa:       0002                    c.slli64        zero
    efac:       00070003                lb      zero,0(a4)
        ...
...
000000000000f030 <camlStr__frametable>:
    f030:       0195                    addi    gp,gp,5
        ...
                        f038: R_RISCV_RELATIVE  *ABS*+0xe668
    f03e:       0000                    unimp
    f040:       00000013                nop
    f044:       0001                    nop
    f046:       0000                    unimp
    f048:       35f0                    fld     fa2,232(a1)
        ...
                        f050: R_RISCV_RELATIVE  *ABS*+0xe618
    f056:       0000                    unimp
    f058:       0011                    c.nop   4
    f05a:       0000                    unimp
    f05c:       35ac                    fld     fa1,104(a1)
        ...
                        f060: R_RISCV_RELATIVE  *ABS*+0xe608
    f066:       0000                    unimp
    f068:       0011                    c.nop   4
    f06a:       0000                    unimp
    f06c:       26fc                    fld     fa5,200(a3)
        ...
...
```
https://github.com/ocaml/ocaml/issues/10688 happens with both gcc 11.2.0 and gcc-11.2.1_p20220115.  But, I heard fedora and debian don't have that issue with same gcc & binutils versions. I have no idea what makes the difference.

gasche commented:
In case someone else doesn't know anything about sections and relocations and stuff, I found some explanations in [this blog post](https://flameeyes.blog/2016/01/16/textrels-text-relocations-and-their-impact-on-hardening-techniques/). What I understood:
- Relocations require an initialization-time mutation of the instruction.
- If they are placed in read-only data sections then the read-onliness cannot easily be enforced by the OS.
-  So "hardened" distributions forbid "text relocations", and force all relocations to belong to "mutable" sections of the object file.

gasche commented:
A naive question from someone that doesn't know anything about these things, sorry.

If I understand correctly, the frametable is supposed to be a "data" section. Indeed, if I go to `otherlibs/str` after a build, and I run

```
../../ocamlopt.opt -nostdlib -I ../../stdlib -absname -w +a-4-9-41-42-44-45-48 \
  -warn-error +A -bin-annot -g -safe-string -strict-sequence -strict-formats  -function-sections \
  -I . -S str.ml strstubs.o -o test
```

(the `str.cmx` compilation command modified to build a full executable) then `str.s` will contain a part that looks like this:

```
camlStr__frametable:
	.quad	338
	.quad	.L907
	.word	35
	.word	0
	.byte	1
	.byte	0
	.align	4
	.long	(.L908 - .) + 0
	.align	8
	.quad	.L906
	.word	33
	.word	0
	.align	4
	.long	(.L909 - .) + 0
[...]
```

On the other hand, the `objdump` command of @alexfanqi generates output about this section that looks like this:

```
0000000000010170 <camlStr__frametable>:
   10170:	52                   	push   %rdx
   10171:	01 00                	add    %eax,(%rax)
   10173:	00 00                	add    %al,(%rax)
   10175:	00 00                	add    %al,(%rax)
   10177:	00 50 af             	add    %dl,-0x51(%rax)
			10178: R_X86_64_RELATIVE	*ABS*+0xaf50
   1017a:	00 00                	add    %al,(%rax)
   1017c:	00 00                	add    %al,(%rax)
   1017e:	00 00                	add    %al,(%rax)
   10180:	23 00                	and    (%rax),%eax
   10182:	00 00                	add    %al,(%rax)
   10184:	01 00                	add    %eax,(%rax)
   10186:	00 00                	add    %al,(%rax)
   10188:	50                   	push   %rax
   10189:	2a 00                	sub    (%rax),%al
   1018b:	00 00                	add    %al,(%rax)
   1018d:	00 00                	add    %al,(%rax)
   1018f:	00 f9                	add    %bh,%cl
			10190: R_X86_64_RELATIVE	*ABS*+0xaef9
   10191:	ae                   	scas   %es:(%rdi),%al
   10192:	00 00                	add    %al,(%rax)
   10194:	00 00                	add    %al,(%rax)
   10196:	00 00                	add    %al,(%rax)
   10198:	21 00                	and    %eax,(%rax)
   1019a:	00 00                	add    %al,(%rax)
   1019c:	fc                   	cld    
   1019d:	29 00                	sub    %eax,(%rax)
   1019f:	00 f0                	add    %dh,%al
			101a0: R_X86_64_RELATIVE	*ABS*+0xaef0
```

My understanding is that while this section is printed as code, it is not actually code. For example, `23` on line 10180 is hex code for the constant word 35 in the third directive of the assembly (not the beginning of an `and` instruction), and `21` on line 10198 corresponds to the constant word 33.

Question: if the interpretation of this section as code is incorrect, is the idea that they contain relocations also incorrect? (I don't see in the original assembly what would generate such relocations, the labels seem to correspond to data in the current module.) In that case, are the TEXTREL warnings erroneous? Is there a way to tell the corresponding tools to interpret this as just data, not code? (Is it not the purpose of `.rodata`?).

alexfanqi commented:
> #10688 happens with both gcc 11.2.0 and gcc-11.2.1_p20220115. But, I heard fedora and debian don't have that issue with same gcc & binutils versions. 

Just realized this issue is discovered on Gentoo because Gentoo compiles gcc with --enable-default-pie.

> if the interpretation of this section as code is incorrect, is the idea that they contain relocations also incorrect?

@gasche I am not an expert of PIC code or ocaml internals. As far as I know, relocations are not necessarily specific to code. If some data references a location/symbol that can be relocated by linker, there could be a relocation generated. A quick example comes to my mind is global variable pointing to a function.
```
int test(int i){
return 0;
} 
static int (*ptr_test)(int) = &test;
```

 In camlStr_frametable, I think <strike>`.quad	.L906` references a symbol and probably results in all these relocations</strike>. As for whether this is correct for ocaml, I am not sure.

xavierleroy commented:
The two sections `.rodata` and `.text` are both in the same segment, but normally `.text` is for code and `.rodata` for data, so the contents of the latter should not be disassembled.

The point remains that for shared objects or position-independent executables, read-only data containing pointers to symbols need relocation at load time and should better be put in a data segment.  macOS has a special `.const_data` section for this purpose (read-only data that needs relocation).

PR #1323 in 2017 "did the right thing" with frametables for other architectures, but that was before the RISC-V port was added.

gasche commented:
Ah indeed, the patch proposed here is an obvious instance of the #1323 changes. So this is clearly an easy fix, and we should consider backporting in 4.14.

nojb commented:
@alexfanqi I think it would be a good idea to add a Changes entry since this issue affects previous versions of the RISC-V backend. Can you add one? Thanks!

alexfanqi committed:
riscv: Generate frametable in data section to improve code relocatability

Similar to what PowerPC and System-Z have done in commit 24980d3fd9848e281761ef2b8fe383e71261789b

With this commit ocamlopt produces .so shared libraries and PIE relocatable
executables that contain no relocations in the text segment.

nojb commented on Changes:
```diff
@@ -127,6 +127,9 @@ Working version
 
 ### Bug fixes:
 
+- #10688: Move frame descriptor table from rodata to data section on riscv.
+  same as ppc and s390x do, ref #1323
```
```suggestion
- #10688: Move frame descriptor table from `rodata` to `data` section on RISC-V.
  Improves support forbuilding DLLs and PIEs. In particular, this applies to all
  binaries in distributions that build PIEs by default (eg Gentoo and Alpine).
  (Alex Fan, review by Gabriel Scherer)
```

nojb commented:
Merged, thanks. @Octachron: should we backport this to 4.14?

gasche commented:
I think we should backport: this is a very safe fix, and we don't know which 5.x release will support RISCV yet, so having a known-good backend in 4.x matters. (The bug fixed here may not show up that much today, or back when the RISCV backend was written, but it will break more and more system configurations as distributions move to PIE/hardening by default.)

Octachron commented:
I agree that an one-line fix for architecture not yet supported on the 5.x branch should be backported to 4.14 (and integrated to the 4.14.0 release in this case).

nojb commented:
Cherry-picked to 4.14 https://github.com/ocaml/ocaml/commit/ddf99786f8229c8282905af5c7c74360d4f2fbd0 (and moved Changes entry to 4.14 section).
